conduit/Data/Conduit/Internal/Conduit.hs

```diff
@@ -812,6 +813,29 @@ ConduitM left0 =$= ConduitM right0 = ConduitM $ \rest ->
   where
 {-# INLINE [1] (=$=) #-}
 
+fuseEither :: Monad m => ConduitM a b m u -> ConduitM b c m d -> ConduitM a c m (Either u d)
+fuseEither (ConduitM left0) (ConduitM right0) = ConduitM $ \rest ->
+    let goRight final left right =
+            case right of
+                HaveOutput p c o  -> HaveOutput (recurse p) (c >> final) o
+                NeedInput rp rc   -> goLeft rp rc final left
+                Done r2           -> PipeM (final >> return (rest (Right r2)))
+                PipeM mp          -> PipeM (liftM recurse mp)
+                Leftover right' i -> goRight final (HaveOutput left final i) right'
+          where
+            recurse = goRight final left
+
+        goLeft rp rc final left =
+            case left of
+                HaveOutput left' final' o -> goRight final' left' (rp o)
+                NeedInput left' lc        -> NeedInput (recurse . left') (recurse . lc)
+                Done r1                   -> PipeM (final >> return (rest (Left r1)))
```

Comment:
You don't want to call `final` here; once upstream terminates, all of its finalizers must already be called.